executor/envd/service/process_service.py

```
@@ -902,3 +982,77 @@ async def _cleanup_finished_process(self, pid: int):
+        self._returncode = self._pty_process.wait(timeout=timeout)
+        return self._returncode
+
+    def terminate(self):
+        """Terminate the process gracefully."""
+        self._pty_process.terminate(force=False)
+
+    def kill(self):
+        """Forcefully kill the process."""
+        self._pty_process.terminate(force=True)
+
+    def send_signal(self, sig):
```

Ruff (0.14.14)

[error] Redefinition of unused `sig` from line 14: `sig` redefined here

(F811)